akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala

```diff
@@ -5,47 +5,137 @@
+   * @return a new behavior with the specified handling appended
+   */
+  def messageEquals(msg: T, handler: FI.Apply[ActorContext[T], Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](messageHandlers = messageHandlers :+ Case[T, T](msg.getClass, Some(_.equals(msg)), (ctx: ActorContext[T], _: T) ⇒ handler.apply(ctx)))
+
+  /**
+   * Add a new case to the signal handling
+   *
+   * @param `type` type of signal to match
+   * @param handler action to apply if the type matches
+   * @tparam M
+   * @return a new behavior with the specified handling appended
+   */
+  def signal[M <: Signal](`type`: Class[M], handler: FI.Apply2[ActorContext[T], M, Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](signalHandlers = signalHandlers :+ Case[T, Signal](`type`, None, (ctx: ActorContext[T], signal: Signal) ⇒ handler.apply(ctx, signal.asInstanceOf[M])))
+
+  /**
+   * Add a new predicated case to the signal handling.
+   *
+   * @param `type` type of signals to match
+   * @param predicate a predicate that will be evaluated on the argument if the type matches
+   * @param handler action to apply if the type matches and the predicate returns true
+   * @tparam M
+   * @return a new behavior with the specified handling appended
+   */
+  def signal[M <: Signal](`type`: Class[M], predicate: FI.TypedPredicate[M], handler: FI.Apply2[ActorContext[T], M, Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](signalHandlers = signalHandlers :+ Case[T, Signal](`type`, Some((t: Signal) ⇒ predicate.defined(t.asInstanceOf[M])), (ctx: ActorContext[T], signal: Signal) ⇒ handler.apply(ctx, signal.asInstanceOf[M])))
+
+  /**
+   * Add a new case to the signal handling without compile time type check.
+   * Should normally not be used, but when matching on class with generic type
+   * argument it can be useful, e.g. <code>GenMsg.class</code> and <code>(ActorContext<Message> ctx, GenMsg&lt;String&gt; list) -> {...}</code>
+   *
+   * @param `type` type of signal to match
+   * @param handler action to apply when the type matches
+   * @return a new behavior with the specified handling appended
+   */
+  def signalUnchecked[M <: Signal](`type`: Class[_ <: Signal], handler: FI.Apply2[ActorContext[T], M, Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](signalHandlers = signalHandlers :+ Case[T, Signal](`type`, None, (ctx: ActorContext[T], signal: Signal) ⇒ handler.apply(ctx, signal.asInstanceOf[M])))
+
+  /**
+   * Add a new case to the signal handling matching equal signals.
+   *
+   * @param msg the signal to compare to
+   * @param handler action to apply when the message matches
+   * @return a new behavior with the specified handling appended
+   */
+  def signalEquals(msg: Signal, handler: FI.Apply[ActorContext[T], Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](signalHandlers = signalHandlers :+ Case[T, Signal](msg.getClass, Some(_.equals(msg)), (ctx: ActorContext[T], _: Signal) ⇒ handler.apply(ctx)))
 
 }
 
 object BehaviorBuilder {
```

Comment:
Hmm right now the generic type of the builder is fixed.

If we'd want to make `object BehaviorBuilder` a common starting point we might be able to start from `BehaviorBuilder[Nothing]` and then potentially widen the generic type at each step... though that might make the types more complicated and compilation error messages when types don't line up less easy to read. Not sure if that'd be worth it.